pkg/ingester/series_map.go

```diff
@@ -2,67 +2,77 @@ package ingester
 
 import (
 	"sync"
+	"sync/atomic"
+	"unsafe"
 
 	"github.com/prometheus/common/model"
 )
 
-// fingerprintSeriesPair pairs a fingerprint with a memorySeries pointer.
-type fingerprintSeriesPair struct {
-	fp     model.Fingerprint
-	series *memorySeries
-}
+const seriesMapShards = 128
 
 // seriesMap maps fingerprints to memory series. All its methods are
 // goroutine-safe. A seriesMap is effectively is a goroutine-safe version of
 // map[model.Fingerprint]*memorySeries.
 type seriesMap struct {
-	mtx sync.RWMutex
+	size   int32
+	shards []shard
+}
+
+type shard struct {
+	mtx sync.Mutex
 	m   map[model.Fingerprint]*memorySeries
+	// Align this struct.
+	pad [cacheLineSize - unsafe.Sizeof(sync.Mutex{}) - unsafe.Sizeof(map[model.Fingerprint]*memorySeries{})]byte
+}
+
+// fingerprintSeriesPair pairs a fingerprint with a memorySeries pointer.
+type fingerprintSeriesPair struct {
+	fp     model.Fingerprint
+	series *memorySeries
 }
 
 // newSeriesMap returns a newly allocated empty seriesMap. To create a seriesMap
 // based on a prefilled map, use an explicit initializer.
 func newSeriesMap() *seriesMap {
-	return &seriesMap{m: make(map[model.Fingerprint]*memorySeries)}
-}
-
-// length returns the number of mappings in the seriesMap.
-func (sm *seriesMap) length() int {
-	sm.mtx.RLock()
-	defer sm.mtx.RUnlock()
-
-	return len(sm.m)
+	shards := make([]shard, seriesMapShards)
+	for i := 0; i < seriesMapShards; i++ {
+		shards[i].m = map[model.Fingerprint]*memorySeries{}
+	}
+	return &seriesMap{
+		shards: shards,
+	}
 }
 
 // get returns a memorySeries for a fingerprint. Return values have the same
 // semantics as the native Go map.
-func (sm *seriesMap) get(fp model.Fingerprint) (s *memorySeries, ok bool) {
-	sm.mtx.RLock()
-	s, ok = sm.m[fp]
-	// Note that the RUnlock is not done via defer for performance reasons.
-	// TODO(beorn7): Once https://github.com/golang/go/issues/14939 is
-	// fixed, revert to the usual defer idiom.
-	sm.mtx.RUnlock()
-	return
+func (sm *seriesMap) get(fp model.Fingerprint) (*memorySeries, bool) {
+	shard := &sm.shards[hashFP(fp)%seriesMapShards]
+	shard.mtx.Lock()
+	ms, ok := shard.m[fp]
+	shard.mtx.Unlock()
+	return ms, ok
 }
 
 // put adds a mapping to the seriesMap. It panics if s == nil.
 func (sm *seriesMap) put(fp model.Fingerprint, s *memorySeries) {
-	sm.mtx.Lock()
-	defer sm.mtx.Unlock()
+	shard := &sm.shards[hashFP(fp)%seriesMapShards]
```

Comment:
This function no longer panics, do we still want it to? Or else update the comment

Reply:
Yeah I don't think we want that; will update comment.